---
title: "Vignette Title"
author: "Vignette Author"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEncoding{UTF-8}
  %\VignetteEngine{knitr::rmarkdown}
editor_options: 
  chunk_output_type: console
---


```{r}
A <- LM(X = c(2,3),C = 2)
B <- LM(X = c(1,"A",4,"A"),C = 2)
res <- combine(B, A, pattern = "A")

vis(res)
new_layout(res)
layout.show(n=4)
```







```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

In this tutorial we are going to explore some of the cool functionalities in `niceRplots`.

`niceRplots` is a collection of plots using base R graphics. It is specifically designed to handle big amount of data. 


```{r}
library(Seurat)
remotes::install_github("niceRplots")
library(niceRplots)
library(igraph)
```

We will be exploring a subset of a PBMC single cell dataset (GEO: GSE149689). Each sample has been subsampled.
We can first download them with the following comands:

```{r}
download.file("https://raw.githubusercontent.com/NBISweden/workshop-scRNAseq/master/labs/data/covid_data_GSE149689/sub/Normal_PBMC_13.h5","~/Downloads/Normal_PBMC_13.h5")
download.file("https://raw.githubusercontent.com/NBISweden/workshop-scRNAseq/master/labs/data/covid_data_GSE149689/sub/nCoV_PBMC_1.h5","~/Downloads/nCoV_PBMC_1.h5")
```


We can first create a Seurat Object to be able to run some quick single cell analysis.

```{r}
control <- Read10X_h5("~/Downloads/Normal_PBMC_13.h5")
control <- CreateSeuratObject(counts = control,
                              project = "control")

covid <- Read10X_h5("~/Downloads/nCoV_PBMC_1.h5")
covid <- CreateSeuratObject(counts = covid,
                            project = "covid")

data <- merge( control , covid )
```

Once the object is created, we can compute some Quality Control steps:

```{r}
data <- PercentageFeatureSet(data, pattern = "^MT-",col.name = "perc_mito")
data <- PercentageFeatureSet(data, pattern = "RP[LS]",col.name = "perc_ribo")
data <- PercentageFeatureSet(data, pattern = "HB[AB]",col.name = "perc_hb")

data <- NormalizeData(data)
data <- CellCycleScoring(data,
                         s.features = casefold(s.genes,upper = T),
                         g2m.features = casefold(g2m.genes,upper = T) )
colnames(data@meta.data)
```

We can also pre-define some criteria for later removing low quality cells

```{r}
data$is_low_quality <- data$perc_mito > 20
```

We can now use the `violins` function to visualise the data. Here we will focus on a common and very important cell quality measure: the percentage of mitocondrial genes.

```{r}
fl <- letters

mypar(2,5,mar=c(4,3,2,1))

violins(data , gene = "perc_mito" , clustering = "orig.ident")
add_letter(fl[1]);fl<-fl[-1]

violins(data , gene = "perc_mito" , clustering = "orig.ident", smooth = .5)
add_letter(fl[1]);fl<-fl[-1]

violins(data , gene = "perc_mito" , clustering = "orig.ident", smooth = .5, plot_points = F )
add_letter(fl[1]);fl<-fl[-1]

violins(data , gene = "perc_mito" , clustering = "orig.ident", smooth = .5, plot_points = F , col=c("pink","cyan"))
add_letter(fl[1]);fl<-fl[-1]

violins(data , gene = "perc_mito" , clustering = "orig.ident", smooth = .5, transparency = 50)
add_letter(fl[1]);fl<-fl[-1]

violins(data , gene = "perc_mito" , clustering = "orig.ident", smooth = .5, transparency = 50 , pt.col = "black" )
add_letter(fl[1]);fl<-fl[-1]

violins(data , gene = "perc_mito" , clustering = "orig.ident", smooth = .5, transparency = 50 , 
        pt.col = c("grey","firebrick4")[data$is_low_quality+1] )
add_letter(fl[1]);fl<-fl[-1]

violins(data , gene = "perc_mito" , clustering = "orig.ident", smooth = .5, srt = 30)
add_letter(fl[1]);fl<-fl[-1]

violins(data , gene = "perc_mito" , clustering = "orig.ident", smooth = .5, srt = 90, ylab = "perc_mito")
add_letter(fl[1]);fl<-fl[-1]

empty_plot()
add_letter(fl[1]);fl<-fl[-1]

```



```{r}
fl <- letters
feats <- c("nCount_RNA","nFeature_RNA","perc_mito","perc_ribo","S.Score","G2M.Score")

mypar(1,4,mar=c(2,6,1,1))
violist( data , genes = feats , clustering = "orig.ident")
add_letter(fl[1]);fl<-fl[-1]

violist( data , genes = feats , clustering = "orig.ident", plot_points = F)
add_letter(fl[1]);fl<-fl[-1]

violist( data , genes = feats , clustering = "orig.ident", transparency = 50 , 
         pt.col = c("grey","firebrick4")[data$is_low_quality+1] )
add_letter(fl[1]);fl<-fl[-1]

```


```{r}
fl <- letters
feats <- c("nCount_RNA","nFeature_RNA","perc_mito","perc_ribo","S.Score","G2M.Score")

mypar(1,4,mar=c(2,6,1,.5))
barlist( data , genes = feats , clustering = "orig.ident" )
add_letter(fl[1]);fl<-fl[-1]

barlist( data , genes = feats , clustering = "orig.ident" , orderby = "perc_mito" )
add_letter(fl[1]);fl<-fl[-1]

barlist( data , genes = feats , clustering = "orig.ident" , orderby = "perc_mito", draw_mean_lines = F )
add_letter(fl[1]);fl<-fl[-1]

barlist( data , genes = feats , clustering = "orig.ident" , orderby = "perc_mito", draw_mean_lines = F,
         col = c("grey","firebrick4")[data$is_low_quality+1] )
add_letter(fl[1]);fl<-fl[-1]

```


And then run the default Seurat pipeline, integration of samples, and finally running PCA and UMAP.

```{r}
data <- FindVariableFeatures(data,nfeatures = 2000)
data <- ScaleData(data)
data <- SplitObject(data, split.by = "orig.ident")
data <- FindIntegrationAnchors(data)
data <- IntegrateData(data)

data <- ScaleData(data,assay = "integrated" )
data <- RunPCA(data, assay = "integrated")
data <- RunUMAP(data, dims = 1:50, assay = "integrated")
```

Next, we can create a graph from the data and separate them into cell clusters.

```{r}
data <- FindNeighbors(data,
                      dims = 1:50,
                      reduction = "pca",
                      assay = "integrated",
                      k.param=10)

data <- FindClusters( data , graph.name = "integrated_snn" , resolution = 0.1)
data <- FindClusters( data , graph.name = "integrated_snn" , resolution = 0.3)

```


```{r}
fl <- letters
cluster_to_compare <- data@meta.data[,c("integrated_snn_res.0.3","integrated_snn_res.0.1")]

mypar(3,4)
plot_sankey( cluster_to_compare , main="res.0.3 vs res.0.1")
add_letter(fl[1]);fl<-fl[-1]

plot_sankey( cluster_to_compare , plot_weights = F , main="don't plot weights")
add_letter(fl[1]);fl<-fl[-1]

plot_sankey( cluster_to_compare , plot_labels = F , main="don't plot class labels")
add_letter(fl[1]);fl<-fl[-1]

plot_sankey( cluster_to_compare , color_by = 1 , main="color by 1st column")
add_letter(fl[1]);fl<-fl[-1]

plot_sankey( cluster_to_compare , color_by = 2 , main="color by 2nd column")
add_letter(fl[1]);fl<-fl[-1]

plot_sankey( cluster_to_compare , use_w1 = T, use_w2 = F , main="no weight on 2nd column")
add_letter(fl[1]);fl<-fl[-1]

plot_sankey( cluster_to_compare , use_w1 = F, use_w2 = T , main="no weight on 1st column")
add_letter(fl[1]);fl<-fl[-1]

plot_sankey( cluster_to_compare , use_w1 = F, use_w2 = F , main="no weight at all")
add_letter(fl[1]);fl<-fl[-1]

plot_sankey( cluster_to_compare ,  gapv = 0 , main="change gaps between")
add_letter(fl[1]);fl<-fl[-1]

plot_sankey( cluster_to_compare ,  gap2v = 0 , main="change gaps within")
add_letter(fl[1]);fl<-fl[-1]

mypal <- c( RColorBrewer::brewer.pal(9,"Set1") , RColorBrewer::brewer.pal(8,"Set2"))
plot_sankey( cluster_to_compare , pal = mypal, main="change color palette")
add_letter(fl[1]);fl<-fl[-1]

```

Having chosen a particular clustering, we can now plot the clusters onto the UMAP embedding.

```{r}
fl <- letters

mypar(3,3,mar=c(3,3,1,1))

plot_meta(data, feat = "seurat_clusters" , frame=F , main = "no frame")
add_letter(fl[1]);fl<-fl[-1]

plot_meta(data, feat = "seurat_clusters" , frame=F ,cex = 1 , main = "increase point size")
add_letter(fl[1]);fl<-fl[-1]

plot_meta(data, feat = "seurat_clusters" , frame=F,cex = .5 , label = T , main = "add label")
add_letter(fl[1]);fl<-fl[-1]

plot_meta(data, feat = "seurat_clusters" , frame=F ,cex = .5 ,  add_lines = T , main = "link cells from clusters")
add_letter(fl[1]);fl<-fl[-1]

names(data@graphs)
plot_meta(data, feat = "seurat_clusters" , frame=F ,cex = .5 ,  main = "add graph (10% of connections)",
          add_graph = "integrated_snn", percent_connections = 0.1)
add_letter(fl[1]);fl<-fl[-1]

mypal <- c( RColorBrewer::brewer.pal(9,"Set1") , RColorBrewer::brewer.pal(8,"Set2"))
plot_meta(data, feat = "seurat_clusters" , frame=F,cex = .5 ,  col = mypal,main = "change color")
add_letter(fl[1]);fl<-fl[-1]
```

We can now compute differentially expressed genes between clusters for visualization.

```{r}
sample_size <- table(data$seurat_clusters)
sample_size[ sample_size > 50 ] <- 50

DGE_cells <- lapply(names(sample_size), function(x){ 
  set.seed(1)
  sample( colnames(data) [ data$seurat_clusters == x ] , size = sample_size[x])
  })
DGE_cells <- unlist(DGE_cells)

DGE_DATA <- data[, DGE_cells]
detable <- FindAllMarkers( DGE_DATA, only.pos = T,max.cells.per.ident = 50,
                          logfc.threshold = .1,assay = "RNA",
                          min.pct = 0.1)
detable <- detable[ detable$p_val < 0.05,  ]
detable$pct.diff <- detable$pct.1 - detable$pct.2
detable$log.pct.diff <- log2(detable$pct.1 / (detable$pct.2+0.01) )
```

A nice way to visualize differentially expressed genes is by cascating plots. `niceRplots` has several ways to achieve this based on your preferences.

```{r}
detable %>% group_by(cluster)  %>% top_n(-10, p_val) %>% top_n(3, pct.diff) -> tops
ord <- factor(sapply(unique(as.character(tops$gene)),function(x){getcluster(DGE_DATA, x, "seurat_clusters")}))
feats <- unique(as.character(tops$gene))[order(as.numeric( as.character(ord) ))]

mypar(1,3,mar=c(2,6,1,.5))
violist(DGE_DATA, genes = feats, 
          clustering = "seurat_clusters", 
          smooth = 1,
          main = "as violins",
          srt = 90)

barlist(DGE_DATA, genes = feats, 
          clustering = "seurat_clusters", 
          main = "as bars",
          srt = 90)

plot_dots(DGE_DATA, 
          genes = feats, 
          clustering = "seurat_clusters", 
          show_grid = T,
          main = "as dots",
          srt = 90)
# mypar()
# plot_heat(DGE_DATA, genes = feats, 
#           clustering = "seurat_clusters", 
#           show_grid = T,
#           main = "top cluster markers",
#           cex.main=1,font.main=1,cex.col = 1,
#           srt = 90,cex.row = 1.1)
```

```{r}
plot_dots(DGE_DATA, genes = feats, 
          clustering = "seurat_clusters", 
          show_grid = T,
          main = "top cluster markers",
          cex.main=1,font.main=1,cex.col = 1,
          srt = 90,cex.row = 1.1 )

```

```{r}
mypar(6,6,mar=c(1,1,1,1))
feats <- unique(as.character(tops$gene))[order(as.numeric( as.character(ord) ))]

for(i in feats[1:36]){
  plot_feat(data, feat = i , assay = "RNA", frame=F , main=i, cex=.5)}
```


```{r}
mypar(3,3,mar=c(1,1,1,3))

plot_feat(data, feat = "GZMA" , assay = "RNA", frame=F , cex = 1)
add_scale_legend(pal = c("grey90",colorRampPalette( c("grey70","navy") )(99) ) ) 

plot_feat(data, feat = "GZMA" , assay = "RNA", frame=F , cex = 1, red = "pca" )
plot_feat(data, feat = "GZMA" , assay = "RNA", frame=F , cex = 1, red = "pca", dims = c(2,4) )


plot_feat(data, feat = "GZMA" , assay = "RNA", frame=F , cex = 1, label = "seurat_clusters")
plot_feat(data, feat = "GZMA" , assay = "RNA", frame=F , cex = 1, col = c("grey90","grey70","orange3","firebrick"))
add_scale_legend(pal = c("grey90",colorRampPalette( c("grey70","orange3","firebrick") )(99) ) ) 


plot_feat(data, feat = "GZMA" , assay = "RNA", frame=F , cex = 1, add_graph = "integrated_snn" , percent_connections = 0.1)
```


















Vignettes are long form documentation commonly included in packages. Because they are part of the distribution of the package, they need to be as compact as possible. The `html_vignette` output type provides a custom style sheet (and tweaks some options) to ensure that the resulting html is as small as possible. The `html_vignette` format:

- Never uses retina figures
- Has a smaller default figure size
- Uses a custom CSS stylesheet instead of the default Twitter Bootstrap style

## Vignette Info

Note the various macros within the `vignette` section of the metadata block above. These are required in order to instruct R how to build the vignette. Note that you should change the `title` field and the `\VignetteIndexEntry` to match the title of your vignette.

## Styles

The `html_vignette` template includes a basic CSS theme. To override this theme you can specify your own CSS in the document metadata as follows:

    output: 
      rmarkdown::html_vignette:
        css: mystyles.css

## Figures

The figure sizes have been customised so that you can easily put two images side-by-side. 

```{r, fig.show='hold'}
plot(1:10)
plot(10:1)
```

You can enable figure captions by `fig_caption: yes` in YAML:

    output:
      rmarkdown::html_vignette:
        fig_caption: yes

Then you can use the chunk option `fig.cap = "Your figure caption."` in **knitr**.

## More Examples

You can write math expressions, e.g. $Y = X\beta + \epsilon$, footnotes^[A footnote here.], and tables, e.g. using `knitr::kable()`.

```{r, echo=FALSE, results='asis'}
knitr::kable(head(mtcars, 10))
```

Also a quote using `>`:

> "He who gives up [code] safety for [code] speed deserves neither."
([via](https://twitter.com/hadleywickham/status/504368538874703872))
